File: src/server/api/go/internal/modules/model/task.go
package model

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
)

type Task struct {
	ID        uuid.UUID `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	SessionID uuid.UUID `gorm:"type:uuid;not null;index:ix_session_session_id;index:ix_session_session_id_task_id,priority:1;index:ix_session_session_id_task_status,priority:1;uniqueIndex:uq_session_id_task_order,priority:1" json:"session_id"`

	TaskOrder      int               `gorm:"not null;uniqueIndex:uq_session_id_task_order,priority:2" json:"task_order"`
	TaskData       datatypes.JSONMap `gorm:"type:jsonb;not null" swaggertype:"object" json:"task_data"`
	TaskStatus     string            `gorm:"type:text;not null;default:'pending';check:task_status IN ('success','failed','running','pending');index:ix_session_session_id_task_status,priority:2" json:"task_status"`
	IsPlanningTask bool              `gorm:"not null;default:false" json:"is_planning_task"`

	CreatedAt time.Time `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt time.Time `gorm:"autoUpdateTime" json:"updated_at"`

	// Task <-> Session
	Session *Session `gorm:"foreignKey:SessionID;references:ID;constraint:OnDelete:CASCADE,OnUpdate:CASCADE;" json:"session"`

	// Task <-> Message (one-to-many)
	Messages []Message `gorm:"constraint:OnDelete:SET NULL,OnUpdate:CASCADE;" json:"messages"`
}

func (Task) TableName() string { return "tasks" }

File: src/server/core/acontext_core/schema/orm/task.py
from dataclasses import dataclass, field
from sqlalchemy import (
    ForeignKey,
    Index,
    Column,
    Integer,
    String,
    Enum,
    text,
    CheckConstraint,
    UniqueConstraint,
    Boolean,
)
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import JSONB, UUID
from typing import TYPE_CHECKING, Optional, List
from .base import ORM_BASE, CommonMixin
from ..session.task import TaskStatus
from ..utils import asUUID

if TYPE_CHECKING:
    from .session import Session
    from .message import Message

# TaskStatusEnum = Enum(TaskStatus, name="task_status_enum", create_type=True)


@ORM_BASE.mapped
@dataclass
class Task(CommonMixin):
    __tablename__ = "tasks"

    __table_args__ = (
        CheckConstraint(
            "task_status IN ('success', 'failed', 'running', 'pending')",
            name="ck_task_status",
        ),
        UniqueConstraint(
            "session_id",
            "task_order",
            name="uq_session_id_task_order",
        ),
        Index("ix_session_session_id", "session_id"),
        Index("ix_session_session_id_task_id", "session_id", "id"),
        Index("ix_session_session_id_task_status", "session_id", "task_status"),
    )

    session_id: asUUID = field(
        metadata={
            "db": Column(
                UUID(as_uuid=True),
                ForeignKey("sessions.id", ondelete="CASCADE"),
                nullable=False,
            )
        }
    )

    task_order: int = field(metadata={"db": Column(Integer, nullable=False)})

    task_data: dict = field(metadata={"db": Column(JSONB, nullable=False)})

    task_status: str = field(
        default="pending",
        metadata={
            "db": Column(
                String,
                nullable=False,
                server_default="pending",
            )
        },
    )

    is_planning_task: bool = field(
        default=False,
        metadata={
            "db": Column(Boolean, nullable=False, default=False, server_default="false")
        },
    )

    space_digested: bool = field(
        default=False,
        metadata={
            "db": Column(Boolean, nullable=False, default=False, server_default="false")
        },
    )

    # Relationships
    messages: List["Message"] = field(
        default_factory=list,
        metadata={"db": relationship("Message", back_populates="task")},
    )

    session: "Session" = field(
        init=False,
        metadata={"db": relationship("Session", back_populates="tasks")},
    )
